sdk/src/vote_program.rs

```diff
@@ -186,25 +186,64 @@ impl VoteState {
     }
 }
 
-// TODO: Deprecate the RegisterAccount instruction and its awkward delegation
-// semantics.
-pub fn register(keyed_accounts: &mut [KeyedAccount]) -> Result<(), ProgramError> {
+pub fn delegate_account(
+    keyed_accounts: &mut [KeyedAccount],
+    node_id: Pubkey,
+) -> Result<(), ProgramError> {
+    if !check_id(&keyed_accounts[0].account.owner) {
+        error!("account[0] is not assigned to the VOTE_PROGRAM");
+        Err(ProgramError::InvalidArgument)?;
+    }
+
+    if keyed_accounts[0].signer_key().is_none() {
+        error!("account[0] should sign the transaction");
+        Err(ProgramError::InvalidArgument)?;
+    }
+
+    let vote_state = VoteState::deserialize(&keyed_accounts[0].account.userdata);
+    if let Ok(mut vote_state) = vote_state {
+        vote_state.delegate_id = node_id;
+        if vote_state.staker_id == Pubkey::default() {
+            vote_state.staker_id = node_id
+        }
+        vote_state.serialize(&mut keyed_accounts[0].account.userdata)?;
+    } else {
+        error!("account[1] does not have enough space assigned");
+        Err(ProgramError::UserdataTooSmall)?;
+    }
+
+    Ok(())
+}
+
+/// Initialize the vote_state for a staking account
+/// Assumes that the account is being init as part of a account creation or balance transfer and
+/// that the transaction must be signed by the staker's keys
+pub fn init_account(keyed_accounts: &mut [KeyedAccount]) -> Result<(), ProgramError> {
     if !check_id(&keyed_accounts[1].account.owner) {
         error!("account[1] is not assigned to the VOTE_PROGRAM");
         Err(ProgramError::InvalidArgument)?;
     }
 
-    // TODO: This assumes keyed_accounts[0] is the SystemInstruction::CreateAccount
-    // that created keyed_accounts[1]. Putting any other signed instruction in
-    // keyed_accounts[0] would allow the owner to highjack the vote account and
-    // insert itself into the leader rotation.
-    let from_id = keyed_accounts[0].signer_key().unwrap();
+    if keyed_accounts[0].signer_key().is_none() {
```

Comment:
No need for a signature here. We just need to check that `vote_state.delegate_id` is `Pubkey::default()` before setting it.

Reply:
Yeah this was just to avoid having to deserialize in case of spam.

Reply:
Let's save that defense for another day. Requiring that signature will mean you need the mint's keypair and a VoteSigner keypair to create and initialize the account.   You should only need the mint's (the funding account).

Reply:
@garious, Requiring accounts[0]'s signature only requires the "staker's" signature. The VoteSigner Keypair doesn't really come into the equation for initialize. 

Effectively `fund_staking_account` will setup 1 transaction with 2 instructions, signed by the staker, where the instructions do the following in sequence:

1. take tokens from the staker's account and put them into the staking account (only pubkey for the staking account is needed)  
2. init the staking account by setting the staker as the delegate 

So, only the staker's keypair is required to sign the tx (init is also enforcing the same for spam :P).